sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/samplers/TraceIdRatioBasedSampler.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2023 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk
+package trace
+package samplers
+
+import org.typelevel.otel4s.sdk.trace.data.LinkData
+import org.typelevel.otel4s.trace.SpanContext
+import org.typelevel.otel4s.trace.SpanKind
+import scodec.bits.ByteVector
+
+/** @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/sdk/#traceidratiobased]]
+  */
+final class TraceIdRatioBasedSampler private (
+    ratio: Double,
+    idUpperBound: Long
+) extends Sampler {
+
+  def shouldSample(
+      parentContext: Option[SpanContext],
+      traceId: ByteVector,
+      name: String,
+      spanKind: SpanKind,
+      attributes: Attributes,
+      parentLinks: List[LinkData]
+  ): SamplingResult =
+    if (math.abs(traceIdRandomPart(traceId)) < idUpperBound)
+      SamplingResult.RecordAndSample
+    else
+      SamplingResult.Drop
+
+  private def traceIdRandomPart(traceId: ByteVector): Long =
+    traceId.drop(8).toLong()
+
+  // format as: 0.000000
+  val description: String = f"TraceIdRatioBased{$ratio%.6f}".replace(",", ".")
+}
+
+private[samplers] object TraceIdRatioBasedSampler {
+
+  /** Creates a new [[TraceIdRatioBasedSampler]] Sampler.
+    *
+    * The ratio of sampling a trace is equal to that of the specified ratio.
+    *
+    * The algorithm used by the Sampler is undefined, notably it may or may not
+    * use parts of the trace ID when generating a sampling decision.
+    *
+    * @param ratio
+    *   the desired ratio of sampling. Must be >= 0 and <= 1.0.
+    */
+  def create(ratio: Double): Sampler = {
+    require(ratio >= 0 && ratio <= 1.0, "ratio must be >= 0 and <= 1.0")
+
+    val idUpperBound =
+      if (ratio == 0.0) Long.MinValue
+      else if (ratio == 1.0) Long.MaxValue
+      else (ratio * Long.MaxValue).toLong
```

Comment:
is this correct? `ratio == 0.5` should yield `0` not `Long.MaxValue/2`, shouldn't it?

Reply:
I'm struggling very hard to figure out how to do the maths without resorting to `BigDecimal`

Reply:
let me try to clarify.

the `Long` value is used to represent a sequence of bits and is in that sense "unsigned", but when doing maths on it, it's signed. so it can have any value from `Long.MinValue` to `Long.MaxValue`. even if the input ratio is `0.0`, the current implementation will still sample any values in `Long.MinValue until 0L`, which is 50% not 0%

Reply:
The `math.abs` is used here, so the distribution of values from `0` to `Long.MaxValue`.

https://github.com/typelevel/otel4s/pull/355/files#diff-fc234a9eef7868569981b776843221255b54d7835308e46da41ab99eb4bc315eR54

Reply:
ah, I see. oops

Reply:
except it will violate the expectations of a ratio of 0.0

Reply:
hm, I didn't think about it that way. I made an issue for myself https://github.com/typelevel/otel4s/issues/404